## Проект: Рынок видеоигр

### Цели и задачи проекта

#### Цель проекта
Основная цель проекта состоит в подготовке данных для написания статьи для привлечения внимания целевой аудитории любителей старых компьютерных игр и популяризации игры "Секреты Темнолесья". Для достижения этой цели планируется провести исследование развития игрово индустрии начала XXI века (2000-2013 годов) с акцентом на жанр RPG. Исследование охватит ключевые тенденции рынка: рост популярности игровых платформ (PlayStation 2, Xbox, Nintendo Wil), динамику продаж игр различных жанров и региональных предпочтений пользователей. Использование исторических данных из открытых источников обеспечит глубину анализа и повысит интерес читателей к истории развития видеоигр, одновременно стимулируя интерес к игре "Секреты Темнолесья".

#### Задачи проекта
1. Отбор и подготовка данных: 
    - Проверка, очистка и систематизировать предоставленные данные о продажах, платформах, жанрах и оценках игр. 
    - Формирование игровой выборки, включив в нее игры за период с 2000 по 2013 год включительно. 
2. Категоризация и ранжирование информации: 
    - Разделение всех игр на три категории на основе пользовательских и экспертных оценок: низкие, средние и высокие.
    - Определение семи самых значимых игровых платформ за исследуемый период по общему количеству выпущенных игр. 
3. Проведение аналитического исследования: 
    - Исследование изменений популярности игровых платформ с течением времени. 
    - Анализ продажи игр в разрезе жанров и выявление особенностей предпочтений игроков в разных регионах мира.
    - Фокусирование на жанре RPG, изучив его место и эфолюцию в игровой индустрии того времени.

### Описание данных

Данные содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта
1. Загрузка и знакомство с данными:
    - Загрузка библиотек и данных датасета.
    - Знакомство с данными и выводы о них. 
2. Проверка ошибок в данных и их предобработка: 
    - Названия (метки) столбцов датафрейма.
    - Типы данных.
    - Наличие пропусков в данных.
    - Явные и неявные дубликаты в данных. 
3. Фильтрация данных.
4. Категоризация данных. 
5. Итоговые выводы.

---

## Загрузка данных и знакомство с ними

*Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.*


In [1]:
# Импорт необходимых библиотек 
import pandas as pd

In [2]:
# Загрузка анализируемого датасета
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

*Познакомьтесь с данными: выведите первые строки и результат метода `info()`.*


In [3]:
# Вывод первых пяти строк датасета 
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [4]:
# Выведем информацию о датасете при помощи метода info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


1. **Объем данных**: 
    - Датасет содержит 16 956 записей об играх и 11 столбцов, что составляет достаточно большой объем данных, который подходит для полноценного анализа.
2. **Соответствие описанию и пропуски**: 
    - В целом данные соответствуют описанию.
    - Однако в данных присутствует достаточно значительное количество пропусков:
        - Критическое количество пропусков (более 40% данных): 
            - `Critic Score`: 8 242 заполненных значений из 16 956 (примерно 51%). Может вызвать большие проблемы, если в анализе необходимо будет использовать оценки критиков. 
        - Существенное количество пропусков: 
            - `Rating`: 10 085 заполненных значений из 16 956 (примерно 40%).
            - `User Score`: 10 152 заполненных значений из 16 956 (примерно 40%).
        - Незначительное количество пропусков: 
            - `Name`: 2 пропущенных значения. 
            - `Genre`: 2 пропущенных значения.
            - `Year of Release`: 275 пропущенных значений (примерно 1,6% пропусков). 
3. **Типы данных**:
    - В датасете есть несоответствия в типах данных, которые могут помешать математическим операциям и анализу в целом: 
        - `EU sales` и `JP sales` имеют тип `object` вместо ожидаемого `float64`. Это говорит о том, что в этих колонках возможно содержатся нечисловые значения. При этом `NA sales` и `Other sales` определились корректно. 
        - `User Score` также определился некорректно, вместо `float64` мы видим `object`. Это также может привести к тому, что будет невохможно корректно провести анализ. 
4. **Другие особенности датасета**:
    - **Названия столбцов**. Для удобства программирования лучше приводить все столбцы к единому "змеиному" стилю, то есть написать названия нижним регистром и с подчеркиваниями. 
    - **`Year of Release` как `float64`**. Поле с годом выпуска представлено как число с плавающей точкой. Хотя логичнее представить данное поле с целочисленным типом данных `int64`. Но этому препятствуют пропуски в данных. 
    - **Неявные пропуски**: В примере данных видно, что для старых игр отсутствуют оценки критиков и пользователей, а также отсутствует рейтинг. Такое произошло из-за того, что в прошлом такие данные могли не собираться по играм. 

---

## Проверка ошибок в данных и их предобработка


### Названия, или метки, столбцов датафрейма

*Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания. Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.*

In [5]:
# Создадим датасет, с которым будем работать в дальнейшем, на основании первоначального 
df_analysed = df.copy()

In [6]:
# Вывод всех столбцов датасета 
df_analysed.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [7]:
# Приведем все названия столбцов к стилю snake case
df_analysed.columns = df_analysed.columns.str.lower().str.replace(' ', '_')
df_analysed.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

In [ ]:
# Создаем копию датасета до преобразования для возможности проверить сделанные изменения после предобработки
tmp = df_analysed.copy() 
len(tmp)

16956

### Типы данных

*Если встречаются некорректные типы данных, предположите их причины. При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.*

In [9]:
# Посмотрим на уникальные значения по полям, в которых встретились некорректные типы данных
unique_eu_sales = df_analysed['eu_sales'].unique()
print('Уникальные значения по полю eu_sales')
print(unique_eu_sales)
unique_jp_sales = df_analysed['jp_sales'].unique()
print('Уникальные значения по полю jp_sales')
print(unique_jp_sales)
unique_user_score = df_analysed['user_score'].unique()
print('Уникальные значения по полю user_score')
print(unique_user_score)
unique_year_of_release = df_analysed['year_of_release'].unique()
print('Уникальные значения по полю year_of_release')
print(unique_year_of_release)

Уникальные значения по полю eu_sales
['28.96' '3.58' '12.76' '10.93' '8.89' '2.26' '9.14' '9.18' '6.94' '0.63'
 '10.95' '7.47' '6.18' '8.03' '4.89' '8.49' '9.09' '0.4' '3.75' '9.2'
 '4.46' '2.71' '3.44' '5.14' '5.49' '3.9' '5.35' '3.17' '5.09' '4.24'
 '5.04' '5.86' '3.68' '4.19' '5.73' '3.59' '4.51' '2.55' '4.02' '4.37'
 '6.31' '3.45' '2.81' '2.85' '3.49' '0.01' '3.35' '2.04' '3.07' '3.87'
 '3.0' '4.82' '3.64' '2.15' '3.69' '2.65' '2.56' '3.11' '3.14' '1.94'
 '1.95' '2.47' '2.28' '3.42' '3.63' '2.36' '1.71' '1.85' '2.79' '1.24'
 '6.12' '1.53' '3.47' '2.24' '5.01' '2.01' '1.72' '2.07' '6.42' '3.86'
 '0.45' '3.48' '1.89' '5.75' '2.17' '1.37' '2.35' '1.18' '2.11' '1.88'
 '2.83' '2.99' '2.89' '3.27' '2.22' '2.14' '1.45' '1.75' '1.04' '1.77'
 '3.02' '2.75' '2.16' '1.9' '2.59' '2.2' '4.3' '0.93' '2.53' '2.52' '1.79'
 '1.3' '2.6' '1.58' '1.2' '1.56' '1.34' '1.26' '0.83' '6.21' '2.8' '1.59'
 '1.73' '4.33' '1.83' '0.0' '2.18' '1.98' '1.47' '0.67' '1.55' '1.91'
 '0.69' '0.6' '1.93' '1.64' '0.55'

**Причины некорректного определения типов данных могут быть следующими**:
1. Поля `eu_sales`, `jp_sales` и `user_score`:
    - **Причина**: Наличие строкового значения `unknown` (в полях `eu_sales` и `jp_sales`) или `tbd` (в поле `user_score`) в данных.
    - Когда в числовом столбце встречается хотя бы одно нечисловое значение, программа автоматически преобразовает весь столбец в тип `object`.
2. Поле `year_of_release`:
    - **Причина**: Наличие пропущенного значения (`nan`).
    - Тип `int64` не поддерживает пропущенные значения, поэтому такие данные автоматически преобразовываются в тип `float64`.

In [10]:
# Проведем преобразования типов данных 
df_analysed['eu_sales'] = pd.to_numeric(df_analysed['eu_sales'], errors = 'coerce') # Из object во float64
df_analysed['jp_sales'] = pd.to_numeric(df_analysed['jp_sales'], errors = 'coerce') # Из object во float64
df_analysed['user_score'] = pd.to_numeric(df_analysed['user_score'], errors = 'coerce') # Из object во float64
df_analysed['year_of_release'] = df_analysed['year_of_release'].astype('Int64') # Из float64 в int64
df_analysed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  Int64  
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.4+ MB


### Наличие пропусков в данных

*Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.*

In [11]:
# Количество пропусков в каждом столбце в абсолютных значениях 
df_analysed.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [12]:
# Количество пропусков в каждом столбце в относительных значениях значениях 
df_analysed.isna().mean()

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64

In [ ]:
# Информация о пропусках
def show_missing_stats(tmp0):
    """
    Функция для отображения статистики пропущенных значений в DataFrame.
    """
    missing_stats = pd.DataFrame({
        'Кол-во пропусков': tmp0.isnull().sum(),
        'Доля пропусков': tmp0.isnull().mean()
    })
    missing_stats = missing_stats[missing_stats['Кол-во пропусков'] > 0]
    
    if missing_stats.empty:
        return "Пропусков в данных нет"
    
    # Форматируем при выводе через Styler
    return (missing_stats.style.format({'Доля пропусков': '{:.4f}'}).background_gradient(cmap='coolwarm'))
show_missing_stats(df_analysed)

,Кол-во пропусков,Доля пропусков
name,2,0.0001
year_of_release,275,0.0162
genre,2,0.0001
eu_sales,6,0.0004
jp_sales,4,0.0002
critic_score,8714,0.5139
user_score,9268,0.5466
rating,6871,0.4052


*Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.*

Как уже было описано выше, пропуски в анализируемом датасете можно разбить на три группы:
- Критическое количество пропусков (более 40% данных): 
    - `critic_score`: 8 242 заполненных значений из 16 956 (примерно 51%). Может вызвать большие проблемы, если в анализе необходимо будет использовать оценки критиков. 
- Существенное количество пропусков: 
    - `rating`: 10 085 заполненных значений из 16 956 (примерно 40%).
    - `user_score`: 10 152 заполненных значений из 16 956 (примерно 40%).
- Незначительное количество пропусков: 
    - `name`: 2 пропущенных значения. 
    - `genre`: 2 пропущенных значения.
    - `year_of_release`: 275 пропущенных значений (примерно 1,6% пропусков). 
Пропуски, которые относятся к последней группе, возможно связаны с ошибками ввода или изначальным отсутствием информации. А пропуски в первых двух группах, как показывают первые строки датасета, характерны для более старых игр, которые были выпущены до внедрения таких метрик. 

**Возможные действия с данными**:
- **Название игры**: Оставляем без изменений, так как на анализ поле `name` не будет особо влиять. 
- **Год выпуска и жанр**: Можно удалить строки с пропусками по полю `year_of_release`, так как количество пропусков в них незначительно. Восстановить данные по другим данным нет возможности. А в самом исследовании мы будем брать период с 2000 по 2013 года. Также можно удалить строки с пустыми значениями по полю `genre`.
- **Данные о продажах**: Можно заполнить пропуски средними значениями, которые будут рассчитываться с учетом платформы, жанра и региона. 
- **Оценки и рейтинги**: Из выборки будут убираться данные об играх, которые были выпущены за пределами анализируемого периода (до 2000 и после 2013 года). А для остальных данных с пропусками можно добавить специальный индикатор.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [14]:
# Удаляем строки с пустыми значениями по полям 'year_of_release' и 'genre'
df_analysed = df_analysed.dropna(subset = ['year_of_release', 'genre'])

In [15]:
# Заполним пропуски в данных по продажам с учетом платформы, жанра и региона (для каждого региона своя функция)
def fill_eu_sales(row):
    if pd.isna(row['eu_sales']):
        group = df_analysed[
            (df_analysed['platform'] == row['platform']) &
            (df_analysed['genre'] == row['genre'])
        ]
        return group['eu_sales'].mean()
    else:
        return row['eu_sales']
def fill_jp_sales(row):
    if pd.isna(row['jp_sales']):
        group = df_analysed[
            (df_analysed['platform'] == row['platform']) &
            (df_analysed['genre'] == row['genre'])
        ]
        return group['jp_sales'].mean()
    else:
        return row['jp_sales']
# Применим функции к датасету 
df_analysed['eu_sales'] = df_analysed.apply(fill_eu_sales, axis = 1)
df_analysed['jp_sales'] = df_analysed.apply(fill_jp_sales, axis = 1)

In [16]:
# Заполним пропуски индикаторами (-1 или N/A)
df_analysed['critic_score'] = df_analysed['critic_score'].fillna(-1)
df_analysed['user_score'] = df_analysed['user_score'].fillna(-1)
df_analysed['rating'] = df_analysed['rating'].fillna('N/A')

In [ ]:
# Посмотрим какие пропуски остались
show_missing_stats(df_analysed)

'Пропусков в данных нет'

### Явные и неявные дубликаты в данных

*Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания. При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.*

In [18]:
# Посмотрим на уникальные значения по полю 'genre'
unique_genre = df_analysed['genre'].unique()
unique_genre

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [19]:
# Чтобы убрать дублирующие записи, необходимо все данные по жанрам привести к нижнему регистру
df_analysed['genre'] = df_analysed['genre'].str.lower()
unique_genre = df_analysed['genre'].unique() # Повторно проверим после того, как привели данные к нижнему регистру
unique_genre

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

In [20]:
# Посмотрим на уникальные значения по полю 'platform'
unique_platform = df_analysed['platform'].unique()
unique_platform

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

`WS` является неизвестной аббревиатурой для консолей. Скорее всего, имелось ввиду название операционной системы Windows, и поэтому мы можем отнести данное название к платформе `PC`.

In [21]:
# Заменим неизвестное наименование платформы на более подходящее 
df_analysed['platform'] = df_analysed['platform'].str.replace('WS', 'PC')

In [22]:
# Посмотрим на уникальные значения по полю 'rating'
unique_rating = df_analysed['rating'].unique()
unique_rating

array(['E', 'N/A', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'],
      dtype=object)

В данных мы видим рейтинг `K-A`, что является устаревшим рейтингом. Необходимо поменять данный рейтинг на новый вариант этого же рейтинга - `E`.

In [23]:
# Заменим устаревшее название рейтинга 'K-A' на новое значение рейтинга 'E'
df_analysed['rating'] = df_analysed['rating'].str.replace('K-A', 'E')

In [24]:
# Посмотрим на уникальные значения по полю 'year_of_release'
unique_year_of_release = df_analysed['year_of_release'].unique()
unique_year_of_release

<IntegerArray>
[2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010, 2013, 2004,
 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014, 1992, 1997, 1993, 1994,
 1982, 2016, 2003, 1986, 2000, 1995, 1991, 1981, 1987, 1980, 1983]
Length: 37, dtype: Int64

Года выпуска выглядят нормальными, никаких преобразований в них не требуется. 

*После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.*

In [25]:
# Проверка на явные дубликаты в данных 
df_duplicate = df_analysed.duplicated().sum()
print(f'Количество явных дубликатов: {df_duplicate}')

Количество явных дубликатов: 235


*Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.*

В датасете на данный момент присутствует 235 дубликатов. Необходимо удалить все дубликаты, оставив только уникальные строки. 

In [26]:
# Удаление дублирующих строк
df_analysed = df_analysed.drop_duplicates()

In [27]:
# Проведем повторную проверку на наличие явных дубликатов в датасете
df_duplicate = df_analysed.duplicated().sum()
print(f'Количество явных дубликатов после удаления: {df_duplicate}')

Количество явных дубликатов после удаления: 0


*В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.*

In [28]:
# Количество строк в первоначальных данных
initial_row_count = df.shape[0]
print(f'Количество строк в первоначальных данных: {initial_row_count}')

Количество строк в первоначальных данных: 16956


In [29]:
# Количество строк в данных на данный момент
final_row_count = df_analysed.shape[0]
print(f'Количество строк в обработанных данных: {final_row_count}')

Количество строк в обработанных данных: 16444


In [30]:
# Посчитаем изменения в абсолютном и относительном значениях
deleted_row_count = initial_row_count - final_row_count
deleted_row_share = round((deleted_row_count / initial_row_count) * 100, 2)
print(f'Количество удаленных строк - {deleted_row_count}, что составляет {deleted_row_share}% от общего количества строк')

Количество удаленных строк - 512, что составляет 3.02% от общего количества строк


In [ ]:
# Проверим сколько удалено строк датасета
a, b = len(tmp), len(df_analysed)
print(" Было строк в исходном датасете", a,
      '\n', "Осталось строк в датасете после обработки", b,
      '\n', "Удалено строк в датасете после обработки", a-b,
      '\n', "Процент потерь", round((a-b)/a*100, 2))

 Было строк в исходном датасете 16956 
 Осталось строк в датасете после обработки 16444 
 Удалено строк в датасете после обработки 512 
 Процент потерь 3.02


*После проведения предобработки данных напишите общий промежуточный вывод.*

На этапе предобработки данных были устранены пропущенные значения и выполнено приведение типов данных к более корректным. Также была произведена нормализация текстовых данных и очистка данных от дубликатов. В результате проведения предобработки данных их объем сократился на **512 строк**, что составлило **3,02%**. Данный показатель является несущественным для обсепечение репрезентативности анализа.

---

## Фильтрация данных

*Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.*

In [32]:
# Создаем и сохраняем новый срез данных в отдельном датасете
df_actual = df_analysed[(df_analysed['year_of_release'] >= 2000) & (df_analysed['year_of_release'] <= 2013)]

In [33]:
# Посмотрим информацию о получившемся срезе данных
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12781 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12781 non-null  object 
 1   platform         12781 non-null  object 
 2   year_of_release  12781 non-null  Int64  
 3   genre            12781 non-null  object 
 4   na_sales         12781 non-null  float64
 5   eu_sales         12781 non-null  float64
 6   jp_sales         12781 non-null  float64
 7   other_sales      12781 non-null  float64
 8   critic_score     12781 non-null  float64
 9   user_score       12781 non-null  float64
 10  rating           12781 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.2+ MB


In [ ]:
# Проверка результата фильтрации
df_actual.year_of_release.sort_values().unique()

<IntegerArray>
[2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012,
 2013]
Length: 14, dtype: Int64

---

## Категоризация данных
    
*Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).*

In [35]:
# Создаем копию датасета
df_actual = df_actual.copy()

In [36]:
# Создание функции для категоризации игр по оценкам пользователей
def category_user_score(user_score):
    if 8 <= user_score < 10:
        return 'Высокая оценка'
    elif 3 <= user_score < 8:
        return 'Средняя оценка'
    elif 0 <= user_score < 3:
        return 'Низкая оценка'
    else:
        return 'Оценка отсутствует'
# Применим созданную функцию к датасету
df_actual['user_score_category'] = df_actual['user_score'].apply(category_user_score)

In [37]:
# Проверим создание новой колонки с категориями по оценкам пользователей
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_score_category
0,Wii Sports,Wii,2006,sports,41.36,28.96,3.77,8.45,76.0,8.0,E,Высокая оценка
2,Mario Kart Wii,Wii,2008,racing,15.68,12.76,3.79,3.29,82.0,8.3,E,Высокая оценка
3,Wii Sports Resort,Wii,2009,sports,15.61,10.93,3.28,2.95,80.0,8.0,E,Высокая оценка
6,New Super Mario Bros.,DS,2006,platform,11.28,9.14,6.50,2.88,89.0,8.5,E,Высокая оценка
7,Wii Play,Wii,2006,misc,13.96,9.18,2.93,2.84,58.0,6.6,E,Средняя оценка


*Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).*

In [38]:
# Создание функции для категоризации игр по оценкам критиков
def category_critic_score(critic_score):
    if 80 <= critic_score < 100:
        return 'Высокая оценка'
    elif 30 <= critic_score < 80:
        return 'Средняя оценка'
    elif 0 <= critic_score < 30:
        return 'Низкая оценка'
    else:
        return 'Оценка отсутствует'
# Применим созданную функцию к датасету
df_actual['critic_score_category'] = df_actual['critic_score'].apply(category_critic_score)

In [39]:
# Проверим создание новой колонки с категориями по оценкам критиков
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_score_category,critic_score_category
0,Wii Sports,Wii,2006,sports,41.36,28.96,3.77,8.45,76.0,8.0,E,Высокая оценка,Средняя оценка
2,Mario Kart Wii,Wii,2008,racing,15.68,12.76,3.79,3.29,82.0,8.3,E,Высокая оценка,Высокая оценка
3,Wii Sports Resort,Wii,2009,sports,15.61,10.93,3.28,2.95,80.0,8.0,E,Высокая оценка,Высокая оценка
6,New Super Mario Bros.,DS,2006,platform,11.28,9.14,6.50,2.88,89.0,8.5,E,Высокая оценка,Высокая оценка
7,Wii Play,Wii,2006,misc,13.96,9.18,2.93,2.84,58.0,6.6,E,Средняя оценка,Средняя оценка


*После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.*

In [40]:
# Количество игр по оценкам пользователей
group_user_score = df_actual.groupby('user_score_category')['name'].count()
group_user_score

user_score_category
Высокая оценка        2286
Низкая оценка          116
Оценка отсутствует    6298
Средняя оценка        4081
Name: name, dtype: int64

In [41]:
# Количество игр по оценкам критиков
group_critic_score = df_actual.groupby('critic_score_category')['name'].count()
group_critic_score

critic_score_category
Высокая оценка        1692
Низкая оценка           55
Оценка отсутствует    5612
Средняя оценка        5422
Name: name, dtype: int64

*Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.*

In [42]:
# Сгруппируем данные по платформам
platform_counts = df_actual.groupby('platform', as_index = False)['name'].count()

In [43]:
# Для репрезентативности переименуем столбцы в полученной таблице
platform_counts.columns = ['Платформа', 'Количество игр']

In [44]:
# Выводим топ-7 платформ по количеству игр 
platform_counts.sort_values(by = 'Количество игр', ascending = False).head(7)

,Платформа,Количество игр
9,PS2,2127
2,DS,2120
14,Wii,1275
12,PSP,1180
16,X360,1121
10,PS3,1087
4,GBA,811


---

## Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

1. **Исходные данные и первоначальная оценка датасета**
    - Для анализа данных был предоставлен обширный датасет, в котором содержится 16 956 записей об играх с 11 характеристиками. Стоит отметить, что данные требовали проработки.
    - Были выявлены следующие проблемы: 
        - **Значительные пропуски**: Данные по оценкам пользователей (`user_score`), критиков (`critic_score`) и возрастным рейтингам ( `rating`) отсутствовало примерно 40 и более процентов записей. Это связано с тем, что для старых игр такие метрики, скорее всего, не собирались. 
        - **Незначительные пропуски**: Небольшое количество пропусков было обнаружено в названиях игр, жанрах и годе выпуска (менее 2%). 
        - **Некорректные типы данных**: Поля с продажами в Европе и Японии, а также с оценкой пользователей были сохранены в типе `object`, что делает невозможным дальнейшее проведение математического анализа без преобразований. 
2. **Предобработка данных**
    - Для обеспечения качества и достоверности анализа был проведен комплекс мер по их предобработке: 
        - **Нормализация**: Названия столбцов были приведены к единому стилю `snake case` (нижний регистр с подчеркиваниями).
        - **Очистка**: Удалены дубликаты и строки с минимальным количеством пропусков (по году выпуска и жанрам). В результате чего первоначальный датасет был сокращен на 512 записей (3,02%), что является несущественной потерей для репрезентативности анализа. 
        - **Причины некоррктного определения типов данных**: 
            - Наличие строковых значений `unknown` и `tbd` в полях `eu_sales`, `jp_sales` и `user_score`.
            - Наличие пропущенных значений в поле `year_of_release`.
        - **Преобразование типов данных**: Исправлены типы данных для колонок с продажами и оценками, что позволило работать с ними как с числовыми данными.
3. **Ключевые результаты анализа за период с 2000 по 2013 года**
    - Анализ был сфокусипован на 2000-2013 годах, которые являются ключевыми дял понимания современного состояния игровой индустрии.
        - **Игровые платформы**: Анализ топ-7 платформ показал, что с большим отрывом в анализируемый период лидируют платформы PS2 и DS. А в сумме 3 из 7 позиций принадлежат платформам компании SONY.
        - **Анализ оценок**: Для анализа были созданы категории по оценкам пользователей и оценкам критиков. Распределение показало: 
            - **Оценки пользователей**: Большинство игр, а именно 4081, поличили средние оценки от пользователей, а высокие оценки получили 2286 игр. Низких оценок значительно меньше - 116.
            - **Оценки критиков**: Аналогичная ситуация наблюдается и среди оценок критиков: средние оценки получили 5422 игры, высокие - 1692 игры, а низкие - 55 игр. 
    - При этом стоит еще отметить, что для старых игр оценки отсутствуют. 